File: training-data/scala/CycleDetection.scala
object CycleDetection {
  def hasCycle(graph: Map[Int, List[Int]]): Boolean = {
    val visited = scala.collection.mutable.Set[Int]()
    val inStack = scala.collection.mutable.Set[Int]()

    def dfs(node: Int): Boolean = {
      visited += node
      inStack += node
      for (neighbor <- graph.getOrElse(node, Nil)) {
        if (!visited.contains(neighbor)) {
          if (dfs(neighbor)) return true
        } else if (inStack.contains(neighbor)) {
          return true
        }
      }
      inStack -= node
      false
    }

    graph.keys.exists(k => !visited.contains(k) && dfs(k))
  }

  def main(args: Array[String]): Unit = {
    val acyclic = Map(0 -> List(1), 1 -> List(2), 2 -> List())
    val cyclic = Map(0 -> List(1), 1 -> List(2), 2 -> List(0))
    println(hasCycle(acyclic))
    println(hasCycle(cyclic))
  }
}

File: training-data/haskell/CycleDetection.hs
import qualified Data.Map as Map
import qualified Data.Set as Set
import Data.Map (Map)
import Data.Set (Set)

type Graph = Map String [String]

hasCycle :: Graph -> Bool
hasCycle g = any (\n -> visit n Set.empty Set.empty) (Map.keys g)
  where
    visit node visited inStack
      | Set.member node inStack = True
      | Set.member node visited = False
      | otherwise =
          let inStack' = Set.insert node inStack
              visited' = Set.insert node visited
          in any (\next -> visit next visited' inStack') (Map.findWithDefault [] node g)

main :: IO ()
main = do
  let acyclic = Map.fromList [("a", ["b"]), ("b", ["c"]), ("c", [])]
      cyclic = Map.fromList [("a", ["b"]), ("b", ["c"]), ("c", ["a"])]
  print (hasCycle acyclic)
  print (hasCycle cyclic)
